#FacialKeypointsDetection

In [1]:
import pandas as pd
import pickle

##数据预处理和探索阶段

In [21]:
Xtrain = pd.read_csv('data/imagesTrain.csv')
train_df = pd.read_csv('data/training.csv')

In [8]:
Xtrain.head()

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0     238     236     237     238     240     240     239     241     241   
1     219     215     204     196     204     211     212     200     180   
2     144     142     159     180     188     188     184     180     167   
3     193     192     193     194     194     194     193     192     168   
4     147     148     160     196     215     214     216     217     219   

   pixel9    ...      pixel9206  pixel9207  pixel9208  pixel9209  pixel9210  \
0     243    ...             33         29         30         34         39   
1     168    ...              1          1          1          1          1   
2     132    ...             64         60         56         61         70   
3     111    ...              1          1          1          1          1   
4     220    ...             33         34         37         37         43   

   pixel9211  pixel9212  pixel9213  pixel9214  pixel9215  
0         49         62         70         75         90  
1          1          1          1          1          1  
2         69         71         78         78         77  
3          1          1          1          1          1  
4         46         83        140        170        176  

[5 rows x 9216 columns]

In [13]:
train_df.head()

left_eye_center_x  left_eye_center_y  right_eye_center_x  \
0          66.033564          39.002274           30.227008   
1          64.332936          34.970077           29.949277   
2          65.057053          34.909642           30.903789   
3          65.225739          37.261774           32.023096   
4          66.725301          39.621261           32.244810   

   right_eye_center_y  left_eye_inner_corner_x  left_eye_inner_corner_y  \
0           36.421678                59.582075                39.647423   
1           33.448715                58.856170                35.274349   
2           34.909642                59.412000                36.320968   
3           37.261774                60.003339                39.127179   
4           38.042032                58.565890                39.621261   

   left_eye_outer_corner_x  left_eye_outer_corner_y  right_eye_inner_corner_x  \
0                73.130346                39.969997                 36.356571   
1                70.722723                36.187166                 36.034723   
2                70.984421                36.320968                 37.678105   
3                72.314713                38.380967                 37.618643   
4                72.515926                39.884466                 36.982380   

   right_eye_inner_corner_y            ...              nose_tip_x  \
0                 37.389402            ...               44.420571   
1                 34.361532            ...               48.206298   
2                 36.320968            ...               47.557263   
3                 38.754115            ...               51.885078   
4                 39.094852            ...               43.299534   

   nose_tip_y  mouth_left_corner_x  mouth_left_corner_y  mouth_right_corner_x  \
0   57.066803            61.195308            79.970165             28.614496   
1   55.660936            56.421447            76.352000             35.122383   
2   53.538947            60.822947            73.014316             33.726316   
3   54.166539            65.598887            72.703722             37.245496   
4   64.889521            60.671411            77.523239             31.191755   

   mouth_right_corner_y  mouth_center_top_lip_x  mouth_center_top_lip_y  \
0             77.388992               43.312602               72.935459   
1             76.047660               46.684596               70.266553   
2             72.732000               47.274947               70.191789   
3             74.195478               50.303165               70.091687   
4             76.997301               44.962748               73.707387   

   mouth_center_bottom_lip_x  mouth_center_bottom_lip_y  
0                  43.130707                  84.485774  
1                  45.467915                  85.480170  
2                  47.274947                  78.659368  
3                  51.561183                  78.268383  
4                  44.227141                  86.871166  

[5 rows x 30 columns]

In [14]:
train_df.describe()

left_eye_center_x  left_eye_center_y  right_eye_center_x  \
count        7039.000000        7039.000000         7036.000000   
mean           66.359021          37.651234           30.306102   
std             3.448233           3.152926            3.083230   
min            22.763345           1.616512            0.686592   
25%            65.082895          35.900451           28.783339   
50%            66.497566          37.528055           30.251378   
75%            68.024752          39.258449           31.768334   
max            94.689280          80.502649           85.039381   

       right_eye_center_y  left_eye_inner_corner_x  left_eye_inner_corner_y  \
count         7036.000000              2271.000000              2271.000000   
mean            37.976943                59.159339                37.944752   
std              3.033621                 2.690354                 2.307332   
min              4.091264                19.064954                27.190098   
25%             36.327681                58.039339                36.626283   
50%             37.813273                59.304615                37.878040   
75%             39.566729                60.519810                39.260348   
max             81.270911                84.440991                66.562559   

       left_eye_outer_corner_x  left_eye_outer_corner_y  \
count              2267.000000              2267.000000   
mean                 73.330478                37.707008   
std                   3.405852                 2.881438   
min                  27.571879                26.250023   
25%                  71.716503                36.085523   
50%                  73.247824                37.639015   
75%                  75.015645                39.373489   
max                  95.258090                64.618230   

       right_eye_inner_corner_x  right_eye_inner_corner_y  \
count               2268.000000               2268.000000   
mean                  36.652607                 37.989902   
std                    2.350268                  2.311907   
min                    5.751046                 26.250023   
25%                   35.506126                 36.766783   
50%                   36.651562                 37.944075   
75%                   37.753571                 39.192917   
max                   70.714966                 69.808803   

                 ...               nose_tip_x   nose_tip_y  \
count            ...              7049.000000  7049.000000   
mean             ...                48.374186    62.715884   
std              ...                 4.280038     5.720567   
min              ...                12.944699    17.932414   
25%              ...                46.602370    59.292697   
50%              ...                48.424713    63.451421   
75%              ...                50.330437    66.492686   
max              ...                89.438592    95.935644   

       mouth_left_corner_x  mouth_left_corner_y  mouth_right_corner_x  \
count          2269.000000          2269.000000           2270.000000   
mean             63.285735            75.970714             32.900404   
std               4.079063             4.579724              3.964552   
min              22.923362            57.023258              2.245766   
25%              61.257143            72.875707             30.797646   
50%              63.178286            75.778689             32.982305   
75%              65.376243            78.879123             35.101302   
max              84.767123            94.673637             74.017993   

       mouth_right_corner_y  mouth_center_top_lip_x  mouth_center_top_lip_y  \
count           2270.000000             2275.000000             2275.000000   
mean              76.179768               47.975412               72.919443   
std                4.417814                3.323060                5.222500   
min               56.690208               12.605173               56.719043 

##数据中含有空值，只有不部分数据预测了某些特征，例如left_eye_inner_corner,只有2271个数据预测了该值，因此需要划分数据集

In [26]:
X=pd.concat([Xtrain,train_df],axis=1)

In [28]:
##保存数据
X.to_pickle('data/Xtrain.pk')

###提取数据，将数据划分为两个部分,同时处理空值，采用fill的方式处理

In [17]:
X= pd.read_pickle('data/Xtrain.pk')

In [18]:
X1 = X[pd.notnull(X['right_eye_inner_corner_x'])]

In [20]:
X1.fillna(X1.mean())
X1.head()

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0     238     236     237     238     240     240     239     241     241   
1     219     215     204     196     204     211     212     200     180   
2     144     142     159     180     188     188     184     180     167   
3     193     192     193     194     194     194     193     192     168   
4     147     148     160     196     215     214     216     217     219   

   pixel9            ...              nose_tip_x  nose_tip_y  \
0     243            ...               44.420571   57.066803   
1     168            ...               48.206298   55.660936   
2     132            ...               47.557263   53.538947   
3     111            ...               51.885078   54.166539   
4     220            ...               43.299534   64.889521   

   mouth_left_corner_x  mouth_left_corner_y  mouth_right_corner_x  \
0            61.195308            79.970165             28.614496   
1            56.421447            76.352000             35.122383   
2            60.822947            73.014316             33.726316   
3            65.598887            72.703722             37.245496   
4            60.671411            77.523239             31.191755   

   mouth_right_corner_y  mouth_center_top_lip_x  mouth_center_top_lip_y  \
0             77.388992               43.312602               72.935459   
1             76.047660               46.684596               70.266553   
2             72.732000               47.274947               70.191789   
3             74.195478               50.303165               70.091687   
4             76.997301               44.962748               73.707387   

   mouth_center_bottom_lip_x  mouth_center_bottom_lip_y  
0                  43.130707                  84.485774  
1                  45.467915                  85.480170  
2                  47.274947                  78.659368  
3                  51.561183                  78.268383  
4                  44.227141                  86.871166  

[5 rows x 9246 columns]

In [22]:
temp = train_df.describe()

In [55]:
#数据对应的列名
cols1 = [i for i in temp.ix['count'].index if temp.ix['count'][i] < 3000]
cols2 = [i for i in temp.ix['count'].index if temp.ix['count'][i] > 3000]

In [47]:
X1.to_pickle('data\X1.pk')

In [48]:
X2 = X[pd.notnull(X[cols2[0]])]

In [54]:
X2.fillna(X2.mean())
X2.to_pickle('data\X2.pk')